File: utils/arn.py
from __future__ import annotations
import re
import fnmatch
from dataclasses import dataclass
from functools import lru_cache, cached_property
from typing import Dict, Optional, Tuple, Iterable
from collections.abc import Mapping, Iterable

_arn_regex = re.compile(
    r"^arn:(?P<partition>[^:]+):(?P<service>[^:]*):(?P<region>[^:]*):"
    r"(?P<account_id>[^:]*):(?P<resource>.+)$"
)


@dataclass(frozen=True, init=False)
class ARN:
    """Immutable AWS ARN representation with ergonomic helpers."""

    raw: str
    partition: str
    service: str
    region: str
    account_id: str
    resource: str

    # ------------------------------------------------------------------
    # Constructors
    # ------------------------------------------------------------------
    def __new__(cls, value: str):
        if not isinstance(value, str):
            raise TypeError("ARN must be constructed from a string")

        value = value.rstrip(":/")
        m = _arn_regex.match(value)
        if not m:
            raise ValueError(f"Invalid ARN: {value}")

        parts = m.groupdict()
        self = super().__new__(cls)
        object.__setattr__(self, "raw", value)
        for key, val in parts.items():
            object.__setattr__(self, key, val)
        return self

    # ------------------------------------------------------------------
    # Robust parsing
    # ------------------------------------------------------------------
    @classmethod
    @lru_cache(maxsize=4096)
    def parse_cached(cls, value: object) -> ARN:
        """
        Cached parser that tolerates ARN strings, dicts, and ARN instances.
        """
        if isinstance(value, ARN):
            return value

        if isinstance(value, dict):
            # Common boto3/Connect patterns
            for key in (
                "Arn",
                "ResourceArn",
                "QueueArn",
                "InstanceArn",
                "ContactFlowArn",
            ):
                if key in value and isinstance(value[key], str):
                    return cls.parse_cached(value[key])
            raise ValueError(f"Unrecognized ARN dict: {value}")

        if not isinstance(value, str):
            raise TypeError(
                f"Expected str or dict for ARN.parse_cached, got {type(value).__name__}"
            )

        if not value.startswith("arn:"):
            raise ValueError(f"Not a valid ARN string: {value}")

        return cls(value)

    @classmethod
    def parse(cls, value: object) -> ARN:
        """
        Non-cached version of parse_cached — same tolerant behavior.
        """
        if isinstance(value, ARN):
            return value

        if isinstance(value, dict):
            for key in (
                "Arn",
                "ResourceArn",
                "QueueArn",
                "InstanceArn",
                "ContactFlowArn",
            ):
                if key in value and isinstance(value[key], str):
                    return cls(value[key])
            raise ValueError(f"Invalid ARN dict: {value}")

        if not isinstance(value, str):
            raise TypeError(
                f"Expected str or dict for ARN.parse, got {type(value).__name__}"
            )

        if not value.startswith("arn:"):
            raise ValueError(f"Not a valid ARN string: {value}")

        return cls(value)

    @classmethod
    def try_parse(cls, value: str) -> Optional[ARN]:
        try:
            return cls.parse(value)
        except Exception:
            return None

    @staticmethod
    def is_valid(value: str) -> bool:
        return bool(_arn_regex.match(value))

    @classmethod
    def from_parts(
        cls,
        service: str,
        resource: str,
        region: str = "",
        account_id: str = "",
        partition: str = "aws",
    ) -> ARN:
        return cls(f"arn:{partition}:{service}:{region}:{account_id}:{resource}")

    # ------------------------------------------------------------------
    # String-like behavior
    # ------------------------------------------------------------------
    def __str__(self) -> str:
        return self.raw

    def __repr__(self) -> str:
        return f"ARN({self.raw!r})"

    def __eq__(self, other: object) -> bool:
        return self.raw == (other.raw if isinstance(other, ARN) else other)

    def __hash__(self) -> int:
        return hash(self.raw)

    def __lt__(self, other) -> bool:
        if not isinstance(other, ARN):
            return NotImplemented
        return (self.service, self.resource_type, self.raw) < (
            other.service,
            other.resource_type,
            other.raw,
        )

    def __contains__(self, item: str) -> bool:
        return item in self.raw

    def __format__(self, spec: str) -> str:
        return format(self.raw, spec)

    def __getattr__(self, name: str):
        """Delegate missing attributes to the underlying string for convenience."""
        return getattr(self.raw, name)

    def __json__(self) -> str:
        return self.raw

    def __iter__(self) -> Iterable[tuple[str, str]]:
        """Allow dict(self) to yield fields."""
        yield from self.to_dict().items()

    def __fspath__(self) -> str:
        return self.raw

    # ------------------------------------------------------------------
    # AWS semantics
    # ------------------------------------------------------------------
    def is_aws_managed(self) -> bool:
        patterns = [
            "arn:aws:iam::aws:policy/*",
            "arn:aws:iam::aws:role/service-role/*",
            "*AWSServiceRoleFor*",
        ]
        return any(fnmatch.fnmatch(self.raw, p) for p in patterns)

    def is_service_linked_role(self) -> bool:
        return self.service == "iam" and "AWSServiceRoleFor" in self.resource

    def is_aws_managed_like(self) -> bool:
        return self.is_aws_managed() or self.is_service_linked_role()

    def matches(self, pattern: str) -> bool:
        """fnmatch convenience (supports wildcards)."""
        return fnmatch.fnmatch(self.raw, pattern)

    def same_resource(self, other: ARN | str) -> bool:
        """True if same service + resource path, ignoring region/account."""
        if not isinstance(other, ARN):
            other = ARN.parse(other)
        return (self.service, self.resource) == (other.service, other.resource)

    # ------------------------------------------------------------------
    # Hierarchy helpers
    # ------------------------------------------------------------------
    def parent(self) -> Optional[ARN]:
        for sep in (":", "/"):
            if sep in self.resource:
                parent_resource = self.resource.rsplit(sep, 1)[0]
                return ARN.from_parts(
                    self.service,
                    parent_resource,
                    self.region,
                    self.account_id,
                    self.partition,
                )
        return None

    @cached_property
    def resource_parts(self) -> list[str]:
        return [p for p in re.split(r"[:/]", self.resource) if p]

    @cached_property
    def resource_type(self) -> str:
        parts = self.resource_parts

        #
        # CONNECT — full fix (handles versions, nested resources, files)
        #
        if self.service == "connect":
            if len(parts) == 2 and parts[0] == "instance":
                return "instance"            
            if len(parts) >= 3 and parts[0] == "instance":
                return parts[2]  # ALWAYS correct primary type
            return "unknown"

        #
        # LAMBDA — functions vs layers vs layer-versions
        #
        if self.service == "lambda":
            if len(parts) >= 1:
                if parts[0] == "function":
                    return "function"
                if parts[0] == "layer":
                    return "layer"
            return "unknown"

        #
        # API GATEWAY — restapis/*/*
        #
        if self.service == "apigateway":
            # Patterns:
            #   restapis/<id>/resources/<id>
            #   restapis/<id>/stages/<stage>
            if len(parts) >= 3 and parts[0] == "restapis":
                return parts[2]
            return "restapi"

        #
        # ELB / ELBv2 — loadbalancer | targetgroup
        #
        if self.service == "elasticloadbalancing":
            if len(parts) >= 1:
                if parts[0] in ("loadbalancer", "targetgroup"):
                    return parts[0]
            return "unknown"

        #
        # IAM — roles, policies, instance-profiles, users, groups
        #
        if self.service == "iam":
            # iam:role/<name>
            # iam:user/<name>
            # iam:policy/<name>
            # iam:instance-profile/<name>
            if len(parts) >= 1:
                return parts[0]
            return "unknown"

        #
        # S3 — bucket vs objects
        #
        if self.service == "s3":
            # arn:aws:s3:::bucket
            # arn:aws:s3:::bucket/object
            return "object" if len(parts) > 1 else "bucket"

        #
        # LEX V2 — handle bots, locales, intents, slots, slot-types, aliases
        #
        if self.service == "lex":
            # Lex has complex multi-layer structure.
            # We always examine even-numbered hierarchy pairs.
            if "bot-alias" in parts:
                return "bot-alias"
            if "bot-locale" in parts:
                return "bot-locale"
            if "intent" in parts:
                return "intent"
            if "slot" in parts:
                return "slot"
            if "slot-type" in parts:
                return "slot-type"
            if parts[:1] == ["bot"]:
                return "bot"
            return "unknown"

        #
        # DEFAULT fallback
        #
        if len(parts) >= 2:
            return parts[-2]
        return parts[0] if parts else "unknown"

    @cached_property
    def resource_id(self) -> str:
        """
        The final identifier for the ARN's primary resource.
        For example:
        - lambda:function/MyFunc → MyFunc
        - connect:instance/X/contact-flow/Y → Y
        - lex:bot/BOT/bot-alias/Alias → Alias
        - s3:::bucket/key → key
        """
        return self.resource_parts[-1] if self.resource_parts else self.resource

    def resource_hierarchy(self) -> list[tuple[str, str]]:
        parts = self.resource_parts
        if len(parts) % 2 != 0:
            parts = parts[1:]
        return list(zip(parts[::2], parts[1::2]))

    def subresource_id(self) -> Optional[str]:
        return self.resource_parts[-1] if len(self.resource_parts) >= 2 else None

    def subresource_parent_id(self, parent_type: str) -> Optional[str]:
        parts = self.resource_parts
        for i, p in enumerate(parts):
            if p == parent_type and i + 1 < len(parts):
                return parts[i + 1]
        return None

    def resource_path(self, sep="/") -> str:
        return sep.join(self.resource_parts)

    def resource_name(self) -> str:
        return self.resource_parts[-1] if self.resource_parts else self.resource

    def short(self) -> str:
        """Compact log form like 'lambda:function/MyFunc'."""
        return f"{self.service}:{self.resource_type}/{self.resource_id}"

    # ------------------------------------------------------------------
    # Transformations
    # ------------------------------------------------------------------
    def arn_for_account(self, account: str) -> ARN:
        return ARN.from_parts(
            self.service, self.resource, self.region, account, self.partition
        )

    def with_region(self, region: str) -> ARN:
        return ARN.from_parts(
            self.service, self.resource, region, self.account_id, self.partition
        )

    # ------------------------------------------------------------------
    # CloudFormation / boto helpers
    # ------------------------------------------------------------------
    def to_boto(self) -> str:
        return self.raw

    def to_cfn_ref(self) -> Dict[str, str]:
        return {"Ref": str(self)}

    def to_cfn_getatt(self, attribute: str = "Arn") -> Dict[str, list]:
        return {"Fn::GetAtt": [str(self), attribute]}

    # ------------------------------------------------------------------
    # S3 / artifact support
    # ------------------------------------------------------------------
    def as_s3_components(self) -> Optional[Tuple[str, Optional[str]]]:
        if self.service != "s3":
            return None
        if "/" in self.resource:
            b, k = self.resource.split("/", 1)
            return b, k
        return self.resource, None

    # ------------------------------------------------------------------
    # Serialization
    # ------------------------------------------------------------------
    def to_dict(self) -> Dict[str, str]:
        return {
            "raw": self.raw,
            "partition": self.partition,
            "service": self.service,
            "region": self.region,
            "account_id": self.account_id,
            "resource": self.resource,
        }

    def as_tuple(self) -> tuple[str, str, str, str, str, str]:
        return (
            self.partition,
            self.service,
            self.region,
            self.account_id,
            self.resource,
            self.raw,
        )

def extract_dependencies(
    obj: object,
    *,
    allow_partial: bool = False,
    service_filter: tuple[str, ...] | None = None,
) -> set[ARN]:
    """
    Recursively extract valid, *resolvable* ARN references from arbitrary
    Python objects.

    FIXED BEHAVIOR:
      - Wildcard ARNs (those containing '*') are excluded.
      - IAM/Cross-service policy patterns like arn:aws:s3:::bucket/* are excluded.
      - CFN stack ARNs are excluded unless explicitly seeded.
      - Keeps same high-performance design as original version.
    """

    found_arns: set[str] = set()

    def is_wildcard_arn(s: str) -> bool:
        # Any '*' inside the resource component makes it unresolvable.
        # Matches IAM policy patterns: arn:aws:s3:::bucket/*, lambda:* etc.
        if "*" not in s:
            return False
        # Never treat wildcard ARNs as resolvable graph dependencies
        return True

    def _walk(value: object):
        if value is None:
            return

        # ---------------------------------------------------------------
        # String fast-path
        # ---------------------------------------------------------------
        if isinstance(value, str):
            if not value.startswith("arn:"):
                return
            if len(value) > 512 or " " in value:
                return

            # Quick wildcard check BEFORE regex parsing
            if is_wildcard_arn(value):
                return

            # Regex parse (fast path)
            m = _arn_regex.match(value)
            if not m:
                if not allow_partial:
                    return
                # fallback partial parsing
                if not value.startswith("arn:"):
                    return

            # Filter by service
            service = (
                m.group("service")
                if m
                else value.split(":")[2]
                if ":" in value
                else ""
            )
            if service_filter and service not in service_filter:
                return

            # Additional suppression of known non-resolvable patterns:
            # CloudFormation stacks often appear in IAM policies
            if service == "cloudformation":
                return

            found_arns.add(value.rstrip(":/"))
            return

        # ---------------------------------------------------------------
        # Dict-like
        # ---------------------------------------------------------------
        if isinstance(value, Mapping):
            for v in value.values():
                _walk(v)
            return

        # ---------------------------------------------------------------
        # Iterable (list, tuple, set)
        # ---------------------------------------------------------------
        if isinstance(value, (list, tuple, set, frozenset)):
            for v in value:
                _walk(v)
            return

        # Scalars ignored

    # Begin walk
    _walk(obj)

    # Convert to ARN objects using cached parser
    return {ARN.parse_cached(v) for v in found_arns}

File: backend/src/arn.ts
const ARN_REGEX =
	/^arn:(?<partition>[^:]+):(?<service>[^:]*):(?<region>[^:]*):(?<account_id>[^:]*):(?<resource>.+)$/;

const ARN_CACHE = new Map<string, ParsedARN>();

export interface ParsedARN {
	raw: string;
	partition: string;
	service: string;
	region: string;
	accountId: string;
	resource: string;
	resourceType: string;
	resourceId: string;
	resourceParts: string[];
}

export function parseARN(value: string): ParsedARN | null {
	const normalized = value.replace(/[/:]+$/, "");

	if (ARN_CACHE.has(normalized)) {
		return ARN_CACHE.get(normalized)!;
	}

	const m = normalized.match(ARN_REGEX);
	if (!m) return null;

	const parts = m.groups!;
	const resourceParts = parts.resource.split(/[:/]/).filter(Boolean);
	const resourceType = extractResourceType(parts.service, resourceParts);
	const resourceId = resourceParts[resourceParts.length - 1] ?? parts.resource;

	const result: ParsedARN = {
		raw: normalized,
		partition: parts.partition,
		service: parts.service,
		region: parts.region,
		accountId: parts.account_id,
		resource: parts.resource,
		resourceType,
		resourceId,
		resourceParts,
	};

	if (ARN_CACHE.size < 4096) {
		ARN_CACHE.set(normalized, result);
	}

	return result;
}

function extractResourceType(service: string, parts: string[]): string {
	if (service === "connect") {
		if (parts.length === 2 && parts[0] === "instance") return "instance";
		if (parts.length >= 3 && parts[0] === "instance") return parts[2];
		return "unknown";
	}

	if (service === "lambda") {
		if (!parts.length) return "unknown";
		if (parts[0] === "function") return "function";
		if (parts[0] === "layer")
			return parts.length >= 3 ? "layerversion" : "layer";
		if (parts[0] === "event-source-mapping") return "eventsourcemapping";
		if (parts[0] === "code-signing-config") return "codesigningconfig";
		if (parts[0] === "runtime") return "runtime";
		return "unknown";
	}

	if (service === "apigateway") {
		if (parts.length >= 3 && parts[0] === "restapis") return parts[2];
		return "restapi";
	}

	if (service === "iam" && parts.length >= 1) return parts[0];

	if (service === "s3") return parts.length > 1 ? "object" : "bucket";

	if (service === "elasticloadbalancing") {
		if (parts[0] === "loadbalancer" || parts[0] === "targetgroup")
			return parts[0];
		return "unknown";
	}

	if (parts.length >= 2) return parts[parts.length - 2];
	return parts[0] ?? "unknown";
}

export function isValidARN(value: string): boolean {
	return ARN_REGEX.test(value.replace(/[/:]+$/, ""));
}

export function tryParseARN(value: string): ParsedARN | null {
	return parseARN(value);
}

export function buildARN(
	service: string,
	resource: string,
	region = "",
	accountId = "",
	partition = "aws",
): string {
	return `arn:${partition}:${service}:${region}:${accountId}:${resource}`;
}

export function isAWSManaged(arn: string): boolean {
	const patterns = [
		"arn:aws:iam::aws:policy/*",
		"arn:aws:iam::aws:role/service-role/*",
		"*AWSServiceRoleFor*",
	];
	const normalized = arn.replace(/[/:]+$/, "");
	return patterns.some((pattern) => fnmatch(normalized, pattern));
}

export function isServiceLinkedRole(arn: string): boolean {
	const parsed = parseARN(arn);
	return (
		parsed?.service === "iam" && parsed.resource.includes("AWSServiceRoleFor")
	);
}

export function remapAccount(
	arn: string,
	targetAccountId: string,
): string | null {
	const parsed = parseARN(arn);
	if (!parsed) return null;
	return buildARN(
		parsed.service,
		parsed.resource,
		parsed.region,
		targetAccountId,
		parsed.partition,
	);
}

export function remapRegion(arn: string, targetRegion: string): string | null {
	const parsed = parseARN(arn);
	if (!parsed) return null;
	return buildARN(
		parsed.service,
		parsed.resource,
		targetRegion,
		parsed.accountId,
		parsed.partition,
	);
}

export function canonicalForGraph(arn: string): string | null {
	const parsed = parseARN(arn);
	if (!parsed) return null;

	if (parsed.resource.includes("*")) {
		if (parsed.service === "s3") {
			const bucket = parsed.resource.split("/", 1)[0];
			if (!bucket || bucket.includes("*")) return null;
			return buildARN(
				parsed.service,
				bucket,
				parsed.region,
				parsed.accountId,
				parsed.partition,
			);
		}
		return null;
	}

	if (parsed.service === "s3" && parsed.resource.includes("/")) {
		const bucket = parsed.resource.split("/", 1)[0];
		return buildARN(
			parsed.service,
			bucket,
			parsed.region,
			parsed.accountId,
			parsed.partition,
		);
	}

	return parsed.raw;
}

export interface ExtractedARNs {
	/** Sanitized copy of the input with ARNs replaced by `__REF_{identifier}__` placeholders. */
	sanitizedData: Record<string, unknown>;
	/** ARN → set of path keys (last JSON segment that led to this ARN). */
	arnPaths: Map<string, Set<string>>;
}

export function extractARNs(
	obj: unknown,
	options: { serviceFilter?: string[]; knownBuckets?: Set<string> } = {},
): ExtractedARNs {
	const arnPaths = new Map<string, Set<string>>();

	function isProbableBucketName(s: string): boolean {
		if (!s || !/^[a-z0-9-]+$/.test(s)) return false;
		if (s.length < 3 || s.length > 63) return false;
		if (s.includes(".") || s.startsWith("arn:")) return false;
		if (!s.includes("-") && !/\d/.test(s)) return false;
		if (
			/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/.test(s)
		)
			return false;
		if (["us-east-1", "us-west-2", "eu-west-1"].includes(s)) return false;
		if (/^\d{4}-\d{2}-\d{2}$/.test(s)) return false;
		return true;
	}

	function inBucketContext(path: string[]): boolean {
		return path.some((p) => p.toLowerCase().includes("bucket"));
	}

	function addReference(arn: string, path: string[]): void {
		const labels = arnPaths.get(arn);
		if (!labels) {
			arnPaths.set(arn, new Set([path.at(-1) ?? ""]));
		} else {
			labels.add(path.at(-1) ?? "");
		}
	}

	function sanitizeValue(value: unknown): unknown {
		if (value == null) return value;

		if (typeof value === "string") {
			// Bucket name detection
			if (
				isProbableBucketName(value) &&
				(options.knownBuckets?.has(value) || inBucketContext([]))
			) {
				if (!options.serviceFilter || options.serviceFilter.includes("s3")) {
					const bucketArn = buildARN("s3", value);
					const canonical = canonicalForGraph(bucketArn);
					if (canonical) return `__REF_${canonical}__`;
				}
				return value;
			}

			// ARN detection
			if (
				!value.startsWith("arn:") ||
				value.length > 512 ||
				value.includes(" ")
			) {
				return value;
			}

			const arn = canonicalForGraph(value);
			if (!arn) return value;

			const parsed = parseARN(arn);
			if (parsed?.service === "lambda" && parsed.resourceType === "runtime") {
				return value;
			}

			if (
				options.serviceFilter &&
				parsed &&
				!options.serviceFilter.includes(parsed.service)
			) {
				return value;
			}

			if (
				parsed?.service === "cloudformation" &&
				!["stack", "stackset", "changeset"].includes(parsed.resourceType)
			) {
				return value;
			}

			return `__REF_${arn}__`;
		}

		if (Array.isArray(value)) {
			return value.map((item) => sanitizeValue(item));
		}

		if (typeof value === "object") {
			const out: Record<string, unknown> = {};
			for (const [k, v] of Object.entries(value)) {
				out[k] = sanitizeValue(v);
			}
			return out;
		}

		return value;
	}

	// Walk to collect ARN paths (for edge labels)
	function walk(value: unknown, path: string[] = []): void {
		if (value == null) return;

		if (typeof value === "string") {
			if (
				isProbableBucketName(value) &&
				(options.knownBuckets?.has(value) || inBucketContext(path))
			) {
				if (!options.serviceFilter || options.serviceFilter.includes("s3")) {
					const bucketArn = buildARN("s3", value);
					const canonical = canonicalForGraph(bucketArn);
					if (canonical) addReference(canonical, path);
				}
				return;
			}

			if (
				!value.startsWith("arn:") ||
				value.length > 512 ||
				value.includes(" ")
			) {
				return;
			}

			const arn = canonicalForGraph(value);
			if (!arn) return;

			const parsed = parseARN(arn);
			if (parsed?.service === "lambda" && parsed.resourceType === "runtime") {
				return;
			}

			if (
				options.serviceFilter &&
				parsed &&
				!options.serviceFilter.includes(parsed.service)
			) {
				return;
			}

			if (
				parsed?.service === "cloudformation" &&
				!["stack", "stackset", "changeset"].includes(parsed.resourceType)
			) {
				return;
			}

			addReference(arn, path);
			return;
		}

		if (typeof value === "object") {
			if (Array.isArray(value)) {
				value.forEach((item, i) => walk(item, [...path, String(i)]));
			} else {
				for (const [k, v] of Object.entries(value)) {
					walk(v, [...path, k]);
				}
			}
		}
	}

	walk(obj);
	const sanitizedData = sanitizeValue(obj) as Record<string, unknown>;
	return { sanitizedData, arnPaths };
}

function fnmatch(value: string, pattern: string): boolean {
	const regex = new RegExp(
		"^" +
			pattern.replace(/[.+?^${}()|[\]\\]/g, "\\$&").replace(/\*/g, ".*") +
			"$",
	);
	return regex.test(value);
}

export function shortARN(arn: string): string {
	const parsed = parseARN(arn);
	if (!parsed) return arn;
	return `${parsed.service}:${parsed.resourceType}/${parsed.resourceId}`;
}
